TOOLS-2412 Add --convertLegacyIndexes flag

tfogo opened the pull request.

xdg commented on mongorestore/options.go:
```diff
@@ -65,21 +65,22 @@ func (*InputOptions) Name() string {
+	DryRunOption                    = "--dryRun"
+	WriteConcernOption              = "--writeConcern"
+	NoIndexRestoreOption            = "--noIndexRestore"
+	IgnoreInvalidIndexOptionsOption = "--ignoreInvalidIndexOptions"
```
I think if this is stripping invalid options, we should rename the option to "--removeInvalidIndexOptions".

xdg replied:
Though if we're also going to fix up the index key document, maybe we need a more inclusive name.  "convertLegacyIndexes"?

xdg commented on mongorestore/options.go:
```diff
@@ -88,19 +89,20 @@ type OutputOptions struct {
-	BulkBufferSize           int    `long:"batchSize" default:"1000" hidden:"true"`
+	WriteConcern              string `long:"writeConcern" value-name:"<write-concern>" default-mask:"-" description:"write concern options e.g. --writeConcern majority, --writeConcern '{w: 3, wtimeout: 500, fsync: true, j: true}'"`
+	NoIndexRestore            bool   `long:"noIndexRestore" description:"don't restore indexes"`
+	IgnoreInvalidIndexOptions bool   `long:"ignoreInvalidIndexOptions" description:"Ignores invalid options in index specifications" hidden:"true"`
```
I think given this could be an issue for anyone restoring to shared, we shouldn't make it hidden.

xdg commented on mongorestore/restore.go:
```diff
@@ -25,6 +25,33 @@ import (
+	"bits": true,
+	"min":  true,
+	"max":  true,
+	"coarsestIndexedLevel": true,
```
This (and finest) aren't in the documented index options.  Can you find out if they would actually appear in an index definition document?

tfogo replied:
It's undocumented but they can appear. I found an old JIRA ticket which suggests Lyft uses them. I'll open a DOCS ticket about it.

xdg commented on mongorestore/restore.go:
```diff
@@ -25,6 +25,33 @@ import (
 
 const insertBufferFactor = 16
 
+var indexOptions = map[string]bool{
```
Please add a comment about which source file this is taken from.  Also, please rename to something like `validIndexOptions` for clarity.

tfogo commented:
Hey @xdg I'm still working on figuring out how indexes used to deal with different edge cases in 3.2. But I thought you might want to just review this as it is currently. Might add some different cases to the switch statement depending on what I find in my investigations.

xdg commented on mongorestore/restore.go:
```diff
@@ -25,24 +25,25 @@ import (
-	"language_override":    true,
-	"ns":                   true,
+// validIndexOptions are taken from https://github.com/mongodb/mongo/blob/master/src/mongo/db/index/index_descriptor.h
+var validIndexOptions = map[string]bool{
```
nit: I suggest sorting the list of options below so the order isn't so random.  (Keeps readers from wondering if there is any meaning to the order.)

xdg commented on mongorestore/restore.go:
```diff
@@ -25,24 +25,25 @@ import (
+	"background":              true,
+	"collation":               true,
+	"default_language":        true,
+	"dropDups":                true,
```
IIUC, `dropDups` isn't currently valid.  I wonder if you need to make this map version-specific as it may change over time. Or populate it with some version-specific data at startup.

xdg commented on mongorestore/restore.go:
```diff
@@ -331,11 +332,29 @@ func (restore *MongoRestore) RestoreIntent(intent *intents.Intent) Result {
+					indexJSON = []byte("<error formatting index>")
+				}
+				indexes[i].Key[j].Value = 1
+				log.Logvf(log.Always, "convertLegacyIndexes: converting the legacy vaule of '%s' from '%s' to '%s' in index '%s'", indexes[i].Key[j].Key, "true", "1", string(indexJSON))
```
Rather than doing this logging element by element, I suggest just flagging if the Key changed and then logging the whole Key document at the end if so.

xdg commented on mongorestore/restore.go:
```diff
@@ -331,11 +332,29 @@ func (restore *MongoRestore) RestoreIntent(intent *intents.Intent) Result {
-func stripInvalidOptions(indexes []IndexDocument) []IndexDocument {
+func convertLegacyIndexes(indexes []IndexDocument) []IndexDocument {
 	for i, index := range indexes {
+		for j, elem := range index.Key {
```
Please factor this out into a `convertLegacyIndexKey` (and later `convertLegacyIndexOptions`).  Otherwise, code complexity is high, hurting readability.  Consider how easy this is to follow (also added an error return -- maybe worth considering):

```
func convertLegacyIndexes(indexes []IndexDocument) ([]IndexDocument, error) {
    var err error
    for i, index := range indexes {
        indexes[i], err = convertLegacyIndexKeys(index)
        if err != nil {
             return nil, err
        }
        indexes[i], err = convertLegacyIndexOptions(index)
        if err != nil {
             return nil, err
        }
    }
    return indexes, nil
}

xdg commented on mongorestore/restore.go:
```diff
@@ -331,11 +332,29 @@ func (restore *MongoRestore) RestoreIntent(intent *intents.Intent) Result {
+					optionsJSON = []byte("<error formatting index>")
+				}
 				delete(indexes[i].Options, key)
+				log.Logvf(log.Always, "convertLegacyIndexes: removing invalid option '%v' from index options '%v'", key, string(optionsJSON))
```
Same.  I suggest saving all the logging for after all options are updated.  Rather than saying "removed this option", you can just say "updated options: Old: <old>; New: <new>".

tfogo commented:
@xdg, I implemented your suggestions and also added logic to deal with index keys that include "", 0, and types other than numbers and strings.

I had to make a function to check if a Decimal128 type is 0. That probably belongs in the driver or in bsonutil. Could you please check it's correct though? I believe the trailing 113 bits should all be zero if the value is 0.

xdg commented on mongorestore/restore.go:
```diff
@@ -332,26 +332,87 @@ func (restore *MongoRestore) RestoreIntent(intent *intents.Intent) Result {
+
+func convertLegacyIndexKeys(index IndexDocument) {
+	var converted bool
+	originalJSONString := createExtJSONString(index.Key)
```
you can move this into the "if converted" stanza, as well, as it's only needed conditionally.

tfogo replied:
But index.Key is a slice and I'm editing the underlying array elements. I think this is less intrusive than making a full copy of the array.

xdg replied:
right. nevermind. :-)

tfogo commented on mongorestore/restore.go:
```diff
@@ -301,6 +332,89 @@ func (restore *MongoRestore) RestoreIntent(intent *intents.Intent) Result {
+}
+
+func createExtJSONString(doc interface{}) string {
+	// by default return "<unable to format document>"" since we don't
```
I made a call here not to throw an error when formatting informational messages. Let me know if you think we should be more strict.

xdg commented on mongorestore/restore.go:
```diff
@@ -332,26 +332,87 @@ func (restore *MongoRestore) RestoreIntent(intent *intents.Intent) Result {
-				log.Logvf(log.Always, "convertLegacyIndexes: removing invalid option '%s' from index '%s' on collection '%s",
-					key, indexes[i].Options["name"], indexes[i].Options["ns"])
-			}
+func isZero(num primitive.Decimal128) bool {
```
I would avoid all this and just compare to `primitive.ParseDecimal128("0")`.  Odds are no-one has ever created an index with a decimal 128 anyway.  (I'm intentionally ignoring that Decimal 128 "0.00" is different than "0" or "-0", etc.)

xdg commented on mongorestore/restore.go:
```diff
@@ -332,26 +332,87 @@ func (restore *MongoRestore) RestoreIntent(intent *intents.Intent) Result {
+func createExtJSONString(doc interface{}) string {
+	// by default return "<unable to format document>"" since we don't
+	// want to throw an error when formatting informational messages.
+	// An error would be inconsequential.
```
👍
